scripts/calculate_udf_formulas.py

```diff
@@ -0,0 +1,474 @@
+        raise e
+    assert type(lh_val) in [float, int, str], (
+        f'Evaluation of "{rh_eval_string}" gave invalid output: "{lh_val}" of type "{type(lh_val)}"'
+    )
+
+    values = [lh_val] + rh_values
+    values_2f = [
+        f"{value:.2f}" if isinstance(value, float) else value for value in values
+    ]
+
+    # Print equations with placeholders and populated values
+    logging.info(f"    Calculation:  {formula_fstring.format(*values_2f)}")
+
+    return lh_val
+
+
+def apply_formula(
+    process: Process, step_type: str, formula_fstring: str, placeholders: list[str]
+) -> None:
+    """This function takes the parsed formula and applies it.
+
+    The application will differ depending on the type of step.
+    """
+
+    overwrite = False if "==" in formula_fstring else True
+
+    # Iterate across artifacts
+    if step_type in ["Standard", "Add Labels"]:
+        for art_tuple in get_art_tuples(process):
+            art_in = art_tuple[0]["uri"]
+            art_out = art_tuple[1]["uri"]
+            logging.info(
+                f"Calculations for input-output '{art_in.name}' ({art_in.id}) --> '{art_out.name}' ({art_out.id})"
+            )
+            try:
+                val = eval_rh(
+                    formula_fstring,
+                    placeholders,
+                    art_in=art_in,
+                    art_out=art_out,
+                    step=process,
+                )
+                assign_val_to_placeholder(
+                    val, placeholders[0], art_in, art_out, process, overwrite
+                )
+            except SkipCalculation:
+                continue
+    elif step_type == "No Outputs":
+        logging.info("Step type: No-output")
+        for art_in in [i for i in process.all_inputs() if i.type == "Analyte"]:
+            logging.info(f"Calculations for input '{art_in.name}' ({art_in.id})")
+            try:
+                val = eval_rh(
+                    formula_fstring,
+                    placeholders,
+                    art_in=art_in,
+                    step=process,
+                )
+                assign_val_to_placeholder(
+                    val, placeholders[0], art_in, process, overwrite
+                )
+            except SkipCalculation:
+                continue
+    else:
+        raise NotImplementedError(f"Step type '{step_type}' not implemented")
+
+
+def get_formulas(step, formula_field):
+    """Extract formulas from a step UDF text field. Skip empty lines and comments."""
+
+    udf_names = [udf[0] for udf in step.udf.items()]
+    formula_field_contents = [
+        step.udf.get(udf_name) for udf_name in udf_names if formula_field in udf_name
+    ]
+    assert len(formula_field_contents) == 1, (
+        f"Searching for '{formula_field}' in step UDFs gave {len(formula_field_contents)} matches."
+    )
+    rows = formula_field_contents[0].split("\n")
+
+    formulas = []
+    for row in rows:
+        if row == "":
```

Comment:
One could maybe check for rows with only whitespace as well. Maybe .strip() could be used?

Reply:
4d27194b3aa9bdd0fcaea29e76e3034c7f6639ba